controllers/base.go

```diff
@@ -26,6 +32,24 @@ func (c *ApiController) GetSessionUser() string {
 		return ""
 	}
 
+	userId, _ := user.(string)
+	if strings.Index(userId, "/") < 0 {
+		return ""
+	}
+
+	userObj := object.GetUser(userId)
+	if userObj == nil {
+		return ""
+	}
+
+	// if user login expired, then clean the session and return an empty string
+	if userObj.SigninExpireTime != "" && util.IsTimeExpired(userObj.SigninExpireTime) {
```

Comment:
What if the user logoff and then login?

What if he has multiple sessions?

Reply:
These codes change the session, and is not that strongly related to accounts.

If a user has many sessions, each of the session is independent. Sessions with "auto sign in" won't expire, and sessions without "auto sign in" will expire.

Reply:
But won't `object.UpdateUserInternal(userId, userObj)` write to DB?

Reply:
@hsluoyz Oh, that's right... I did not take this into consideration.

Can I add a time stamp to the session, so the session will become something like this: `built-in/admin 1626358811`

Then add some code to `GetSessionUser` to recognize the time stamp.

Reply:
@hsluoyz Resolved that. Now, exipre timestamps are stored in session.

Checked auto signin:
![](https://user-images.githubusercontent.com/42638489/125813898-e9f7ddec-b75b-4612-9e40-bcac56c145e0.png)

Did not check:
![$( _LRYFMW4ACOUVB4OES5N](https://user-images.githubusercontent.com/42638489/125813961-126daff0-5eaf-41d5-845c-47f65b6e771d.png)

Reply:
@Kininaru  We can store a struct in the session, you can see Casnode's code. We don't need to put all stuff in one string and then parse it. This code is fragile..

Reply:
@hsluoyz Can I add another session called `ExpireTime`? Or store a struct in a session called `SessionInfo`...
I found that many places use the session `username` like `built-in/admin`, turning the session to a struct (or even just add a timestamp behind user id) may cause many unexpected bugs...

Reply:
@Kininaru we can extract `GetSessionUsername()` first before this PR like this:

https://github.com/casbin/casnode/blob/9c66ea1de047a22c9d99e36bcd86dbc9280acb5b/controllers/base.go#L54-L60

Then we can easily change the underlying implementation of session data structure.

Reply:
@hsluoyz Finished. Now, for each session, there is a struct called `SessionData` stored in session like this: 
```go
type SessionData struct {
	ExpireTime int64
}
```
We can add more info to the struct in the future.